presto-memory-context/src/main/java/com/facebook/presto/memory/SimpleLocalMemoryContext.java

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import static com.facebook.presto.memory.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class SimpleLocalMemoryContext
+        implements LocalMemoryContext
+{
+    @Nonnull
+    private final AggregatedMemoryContext parentMemoryContext;
+    @GuardedBy("this")
+    private long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public SimpleLocalMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext);
+    }
+
+    @Override
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    @Override
+    public synchronized ListenableFuture<?> setBytes(long bytes)
+    {
+        checkClosed();
+
+        if (bytes == usedBytes) {
+            return NOT_BLOCKED;
+        }
+
+        // update the parent first as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+        ListenableFuture<?> future = parentMemoryContext.updateBytes(bytes - usedBytes);
+        usedBytes = bytes;
+        return future;
+    }
+
+    @Override
+    public synchronized boolean tryUpdateBytes(long bytes)
+    {
+        checkClosed();
+
+        if (parentMemoryContext.tryUpdateBytes(bytes)) {
+            usedBytes += bytes;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * This method transfers ownership of allocations from this memory context to the "to" memory context,
+     * where parent of this is a descendant of to.parent (there can be multiple AggregatedMemoryContexts between them).
+     *
+     * During the transfer we don't call the reservation handlers to reflect any state changes outside of the contexts.
+     */
+    @Override
+    public synchronized void transferOwnership(LocalMemoryContext to)
+    {
+        checkArgument(to instanceof SimpleLocalMemoryContext, "to has to be SimpleLocalMemoryContext");
+        checkClosed();
+
+        SimpleLocalMemoryContext target = (SimpleLocalMemoryContext) to;
+        AggregatedMemoryContext parent = parentMemoryContext;
+        while (parent != null && parent != target.parentMemoryContext) {
+            synchronized (parent) {
+                parent.usedBytes -= usedBytes;
+            }
+            parent = parent.parentMemoryContext;
+        }
+
```

Comment:
Fail if `parent == null`. This serves 3 purposes:

1. Sanity check
2. Make it clear to readers of this code that the memory is left in a corrupt/inconsistent state at that point.
3. Avoid potential dead lock

1 would be apparent when you add the check. 2 and 3 are probably worth being pointed out explicitly. Put 2 in the error message (or comment, up to you). Put 3 in comment.